gateway-service/src/main/java/org/java/mentorship/gateway/exception/domain/GatewayException.java

```diff
@@ -2,17 +2,31 @@
 
 import lombok.Data;
 import lombok.EqualsAndHashCode;
+import org.java.mentorship.contracts.common.dto.ErrorResponse;
 import org.springframework.http.HttpStatus;
 
+import java.util.Optional;
+
 @EqualsAndHashCode(callSuper = true)
 @Data
 public class GatewayException extends RuntimeException {
 
     private final HttpStatus statusCode;
+    private final ErrorResponse errorResponse;
+    private final Optional<RuntimeException> sourceException;
+
+    public GatewayException(ErrorResponse errorResponse, HttpStatus statusCode) {
+        super(errorResponse.getError());
+        this.statusCode = statusCode;
+        this.errorResponse = errorResponse;
+        this.sourceException = Optional.empty();
+    }
 
-    public GatewayException(String message, HttpStatus statusCode) {
-        super(message);
+    public GatewayException(ErrorResponse errorResponse, RuntimeException sourceException, HttpStatus statusCode) {
+        super(errorResponse.getError(), sourceException);
         this.statusCode = statusCode;
+        this.errorResponse = errorResponse;
+        this.sourceException = Optional.of(sourceException);
```

Comment:
No need for the sourceException local variable.
The parent class holds it and we won't be using it anythime soon.
It's only use would be during logging, i.e. log.error("Message", e)
This would print the full stack trace of the cause